src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -128,6 +133,91 @@ internal partial class Binder
+                ? BindInitializerExpression(syntax: node.Initializer, type: strippedType, typeSyntax: syntax, diagnostics)
+                : null;
+
+            BoundExpression operand = source;
+
+            switch (strippedType.TypeKind)
+            {
+                case TypeKind.Array:
+                case TypeKind.Enum:
+                case TypeKind.Delegate:
+                case TypeKind.Interface:
+                    Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, strippedType);
+                    break;
+
+                case TypeKind.Pointer:
+                    Error(diagnostics, ErrorCode.ERR_UnsafeTypeInObjectCreation, syntax, strippedType);
+                    break;
+
+                case TypeKind.Dynamic:
+                    Error(diagnostics, ErrorCode.ERR_NoConstructors, syntax, strippedType);
+                    break;
+
+                case TypeKind.Struct when strippedType.IsTupleType:
+                    Error(diagnostics, ErrorCode.ERR_NewWithTupleTypeSyntax, syntax, strippedType);
+                    break;
+
+                case TypeKind.Struct:
+                case TypeKind.Class:
+                    operand = BindClassCreationExpression(
+                        node,
+                        typeName: strippedType.Name,
+                        typeNode: node,
+                        type: (NamedTypeSymbol)strippedType,
+                        arguments,
+                        diagnostics,
+                        boundInitializerOpt,
+                        forTargetTypedNew: true);
+                    operand.WasCompilerGenerated = true;
+                    break;
+
+                case TypeKind.TypeParameter:
+                    var typeParameter = (TypeParameterSymbol)strippedType;
+                    if (!typeParameter.IsInstantiable())
+                    {
+                        Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, strippedType);
+                        break;
+                    }
+
+                    operand = BindTypeParameterCreationExpression(
+                        node,
+                        typeParameter,
+                        boundInitializerOpt,
+                        diagnostics);
+                    operand.WasCompilerGenerated = true;
+                    break;
+
+                default:
+                    break;
+            }
+
+            arguments.Free();
+
+            return new BoundConversion(
+                syntax,
+                operand,
+                conversion: destination.IsNullableType()
+                    ? new Conversion(ConversionKind.ImplicitNullable, ImmutableArray.Create(conversion))
+                    : conversion,
+                @checked: false,
+                explicitCastInCode: isCast,
+                constantValueOpt: null, // A "target-typed new" would never produce a constant.
+                type: destination,
+                hasErrors: operand.HasErrors)
```

Comment:
I'm not sure whether or not we should set `hasErrors` here, there are three scenarios that we might want to make it so: (1) If the type is filtered, (2) If the type is `TypeKind.Error`, or anything except those we handled, or (3) If the constructor binds with error.